roachtest: sql: index backfill failed due to duplicate key value

roachtest.sqlsmith/setup=tpcc/setting=ddl-nodrop [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/6698314?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/6698314?buildTab=artifacts#/sqlsmith/setup=tpcc/setting=ddl-nodrop) on master @ [4780ee15194e7d22b661ac92254b13cd2b71dcb1](https://github.com/cockroachdb/cockroach/commits/4780ee15194e7d22b661ac92254b13cd2b71dcb1):


```
test artifacts and logs in: /artifacts/sqlsmith/setup=tpcc/setting=ddl-nodrop/run_1
	sqlsmith.go:304,sqlsmith.go:327,test_runner.go:928: ping node 1: dial tcp 34.139.227.18:26257: connect: connection refused
		HINT: node likely crashed, check logs in artifacts > logs/1.unredacted
		previous sql:
		ALTER TABLE defaultdb.public.warehouse
			ADD COLUMN
				col_11
					GEOMETRY
					DEFAULT (
						st_mpointfromtext('I':::STRING::STRING, crdb_internal.lease_holder('\x49':::BYTES::BYTES)::INT8::INT8)::GEOMETRY
					)
					NOT NULL;ping node 1: dial tcp 34.139.227.18:26257: connect: connection refused
		HINT: node likely crashed, check logs in artifacts > logs/1.unredacted
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=tpcc/setting=ddl-nodrop.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-20078

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x174fc39]

goroutine 7771 [running]:
panic({0x4a8cec0, 0x9367ca0})
	GOROOT/src/runtime/panic.go:987 +0x3ba fp=0xc004f7c278 sp=0xc004f7c1b8 pc=0x48d75a
runtime.panicmem(...)
	GOROOT/src/runtime/panic.go:260
runtime.sigpanic()
	GOROOT/src/runtime/signal_unix.go:835 +0x2f6 fp=0xc004f7c2c8 sp=0xc004f7c278 pc=0x4a45f6
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).Send(0xc004441380?, {_, _}, {{{0x0, 0x0, 0x0}, 0x0, {0x0, 0x0, 0x0}, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/txn.go:1070 +0x39 fp=0xc004f7c590 sp=0xc004f7c2c8 pc=0x174fc39
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).Send-fm({0x6691b20?, _}, {{{0x0, 0x0, 0x0}, 0x0, {0x0, 0x0, 0x0}, {0x0, ...}, ...}, ...})
	<autogenerated>:1 +0x85 fp=0xc004f7c6a0 sp=0xc004f7c590 pc=0x1754b25
github.com/cockroachdb/cockroach/pkg/kv.sendAndFill({0x6691b20, 0xc0045fe2c0}, 0x0?, 0xc0064fcb00)
	github.com/cockroachdb/cockroach/pkg/kv/db.go:844 +0x119 fp=0xc004f7c898 sp=0xc004f7c6a0 pc=0x1746239
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).Run(0x0, {0x6691b20, 0xc0045fe2c0}, 0x18d4527?)
	github.com/cockroachdb/cockroach/pkg/kv/txn.go:673 +0x74 fp=0xc004f7c8e0 sp=0xc004f7c898 pc=0x174d7b4
github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.glob..func242(0xc006ff4f00, {0xc0007fbec0?, 0x0?, 0x3e9?})
	github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/builtins.go:5180 +0x11d fp=0xc004f7c960 sp=0xc004f7c8e0 pc=0x213ac9d
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalFuncExpr(0xc006ff4f00, 0xc0053c2f00)
	github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:459 +0xd6 fp=0xc004f7cbf8 sp=0xc004f7c960 pc=0x17b6cd6
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval(0x6691bc8?, {0x671f520?, 0xc006ff4f00?})
	github.com/cockroachdb/cockroach/bazel-out/k8-opt/bin/pkg/sql/sem/tree/eval_expr_generated.go:264 +0x2d fp=0xc004f7cc18 sp=0xc004f7cbf8 pc=0x14123ed
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).evalFuncArgs(0x0?, 0xc0053c3080)
	github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:476 +0xbf fp=0xc004f7cc88 sp=0xc004f7cc18 pc=0x17b6f7f
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalFuncExpr(0xc006ff4f00, 0xc0053c3080)
	github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:451 +0x90 fp=0xc004f7cf20 sp=0xc004f7cc88 pc=0x17b6c90
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval(0x4a37b60?, {0x671f520?, 0xc006ff4f00?})
	github.com/cockroachdb/cockroach/bazel-out/k8-opt/bin/pkg/sql/sem/tree/eval_expr_generated.go:264 +0x2d fp=0xc004f7cf40 sp=0xc004f7cf20 pc=0x14123ed
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.Expr(...)
	github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:24
github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfiller).BuildIndexEntriesChunk.func1({0xc0007fbb30, 0x1, 0x0?})
	github.com/cockroachdb/cockroach/pkg/sql/backfill/backfill.go:876 +0x102 fp=0xc004f7d000 sp=0xc004f7cf40 pc=0x22f6d42
github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfiller).BuildIndexEntriesChunk(0xc0018b0000, {0x6691bc8, 0xc007920150}, 0xc004421550, {0x6749458, 0xc001c7c000}, {{0xc0084bc9c0, 0x3, 0x8}, {0xc0084bc318, ...}}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/backfill/backfill.go:909 +0xbf3 fp=0xc004f7d568 sp=0xc004f7d000 pc=0x22f5af3
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*indexBackfiller).buildIndexEntryBatch.func1({0x6691bc8, 0xc007920150}, 0x66d8120?)
	github.com/cockroachdb/cockroach/pkg/sql/rowexec/indexbackfiller.go:435 +0xf2 fp=0xc004f7d618 sp=0xc004f7d568 pc=0x252b972
github.com/cockroachdb/cockroach/pkg/kv.runTxn.func1({0x6691bc8?, 0xc007920150?}, 0xc0064fc950?)
	github.com/cockroachdb/cockroach/pkg/kv/db.go:964 +0x27 fp=0xc004f7d640 sp=0xc004f7d618 pc=0x1746967
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc004421550, {0x6691bc8, 0xc007920150}, 0xc004f7d980)
	github.com/cockroachdb/cockroach/pkg/kv/txn.go:961 +0xae fp=0xc004f7d928 sp=0xc004f7d640 pc=0x174f24e
github.com/cockroachdb/cockroach/pkg/kv.runTxn({0x6691bc8, 0xc007920150}, 0x7?, 0x0?)
	github.com/cockroachdb/cockroach/pkg/kv/db.go:963 +0x5a fp=0xc004f7d9a0 sp=0xc004f7d928 pc=0x174687a
github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl(0x6691bc8?, {0x6691bc8, 0xc007920150}, 0xa?, 0x0?, 0x6691bc8?)
	github.com/cockroachdb/cockroach/pkg/kv/db.go:926 +0x89 fp=0xc004f7d9e8 sp=0xc004f7d9a0 pc=0x1746609
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(...)
	github.com/cockroachdb/cockroach/pkg/kv/db.go:905
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*indexBackfiller).buildIndexEntryBatch(0xc0018b0000?, {0x6691bc8?, 0xc0079200f0?}, {{0xc0084bc9c0, 0x3, 0x8}, {0xc0084bc318, 0x3, 0x8}}, {0x17196b9a5d88707c, ...})
	github.com/cockroachdb/cockroach/pkg/sql/rowexec/indexbackfiller.go:428 +0x339 fp=0xc004f7dbb0 sp=0xc004f7d9e8 pc=0x252b4b9
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*indexBackfiller).constructIndexEntries(0xc0018b0000, {0x6691bc8, 0xc0079200f0}, 0xc001b29860)
	github.com/cockroachdb/cockroach/pkg/sql/rowexec/indexbackfiller.go:134 +0x2f8 fp=0xc004f7dec8 sp=0xc004f7dbb0 pc=0x25283f8
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*indexBackfiller).runBackfill.func1({0x6691b20?, 0xc0072b9780?})
	github.com/cockroachdb/cockroach/pkg/sql/rowexec/indexbackfiller.go:319 +0xcb fp=0xc004f7df58 sp=0xc004f7dec8 pc=0x252a1eb
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x25 fp=0xc004f7df78 sp=0xc004f7df58 pc=0x114b0a5
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:74 +0x64 fp=0xc004f7dfe0 sp=0xc004f7df78 pc=0x114a804
runtime.goexit()
	GOROOT/src/runtime/asm_amd64.s:1594 +0x1 fp=0xc004f7dfe8 sp=0xc004f7dfe0 pc=0x4c23c1
created by golang.org/x/sync/errgroup.(*Group).Go
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:71 +0xa5
```

--- a/pkg/sql/logictest/testdata/logic_test/alter_table
+++ b/pkg/sql/logictest/testdata/logic_test/alter_table
@@ -2673,3 +2673,17 @@ CREATE TABLE t1_non_indexable (n INT8);
 
 statement error pq: unimplemented: column x is of type char\[\] and thus is not indexable
 ALTER TABLE t1_non_indexable ADD COLUMN x CHAR(256)[] UNIQUE;
+
+subtest regression_89025
+
+statement ok
+CREATE TABLE t_89025 (i int primary key);
+
+statement ok
+INSERT INTO t_89025 VALUES (1)
+
+statement error pgcode 0A000 cannot use crdb_internal\.lease_holder in this context
+ALTER TABLE t_89025 ADD COLUMN j INT DEFAULT (crdb_internal.lease_holder('a'));
+
+statement ok
+DROP TABLE t_89025

--- a/pkg/sql/sem/builtins/builtins.go
+++ b/pkg/sql/sem/builtins/builtins.go
@@ -5177,6 +5177,10 @@ value if you rely on the HLC for accuracy.`,
 						Key: key,
 					},
 				})
+				if ctx.Txn == nil { // can occur during backfills
+					return nil, pgerror.Newf(pgcode.FeatureNotSupported,
+						"cannot use crdb_internal.lease_holder in this context")
+				}
 				if err := ctx.Txn.Run(ctx.Context, b); err != nil {
 					return nil, pgerror.Wrap(err, pgcode.InvalidParameterValue, "error fetching leaseholder")
 				}

